src/modules/esl-popup/core/esl-popup-position.ts

```diff
@@ -141,19 +143,51 @@ function fitOnMajorAxis(cfg: PopupPositionConfig, value: PopupPositionValue): Po
  * @returns updated popup position value
  * */
 function adjustAlongMajorAxis(cfg: PopupPositionConfig, value: PopupPositionValue): PopupPositionValue {
-  let {popup, placedAt} = value;
+  const popup = isStartingSide(cfg.position)
+    ? adjustForStartingSide(cfg, value.popup)
+    : adjustForEndingSide(cfg, value.popup);
+  const placedAt = getOppositePosition(cfg.position);
+  return {...value, popup, placedAt};
+}
+
+/**
+ * Updates popup rect to fit on major axis in case positioning on the starting side.
+ * @param cfg - popup position config
+ * @param popup - popup rect
+ * @returns updated popup rect
+ * */
+function adjustForStartingSide(cfg: PopupPositionConfig, popup: Rect): Rect {
+  const {position, inner, hasInnerOrigin} = cfg;
   let {x, y} = popup;
-  if (isStartingSide(cfg.position)) {
-    x = cfg.position === 'left' ? cfg.inner.right : x;
-    y = cfg.position === 'top' ? cfg.inner.bottom : y;
-  } else {
-    x = cfg.position === 'right' ? cfg.inner.x - popup.width : x;
-    y = cfg.position === 'bottom' ? cfg.inner.y - popup.height : y;
+  switch (position) {
+    case 'left':
+      x = hasInnerOrigin ? inner.x : inner.right;
+      break;
+    case 'top':
+      y = hasInnerOrigin ? inner.y : inner.bottom;
+      break;
   }
-  popup = new Rect(x, y, popup.width, popup.height);
-  placedAt = getOppositePosition(cfg.position);
+  return new Rect(x, y, popup.width, popup.height);
+}
 
-  return {...value, popup, placedAt};
+/**
+ * Updates popup rect to fit on major axis in case positioning on the ending side.
+ * @param cfg - popup position config
+ * @param popup - popup rect
+ * @returns updated popup rect
+ * */
+function adjustForEndingSide(cfg: PopupPositionConfig, popup: Rect): Rect {
+  const {position, inner, hasInnerOrigin} = cfg;
+  let {x, y} = popup;
+  switch (position) {
+    case 'right':
+      x = (hasInnerOrigin ? inner.right : inner.x) - popup.width;
+      break;
+    case 'bottom':
+      y = (hasInnerOrigin ? inner.bottom : inner.y) - popup.height;
+      break;
+  }
+  return new Rect(x, y, popup.width, popup.height);
 }
```

Comment:
Methods are absolutely the same except invertion of `hasInnerOrigin` input value